spec/users_controller_spec.rb

```diff
@@ -0,0 +1,111 @@
+require_relative 'spec_helper'
+
+def app
+  UsersController.new
+end
+
+describe UsersController do
+  before do
+    @users = DB[:users]
+  end
+
+  after do
+    @users.truncate
+  end
+ 
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as an admin" do
+    before do
+      # DB.transaction(:rollback => :always) do
+        @users.insert(admin: true, fname: "f1", lname: "l1", email: "e1", username: "u1")
+        @users.insert(fname: "different_user", lname: "l1")
+        @admin = @users.where(admin: true).first
+        @user = @users.where(fname: "different_user").first
+        env "rack.session", {:current_user_id => @admin[:id]}
+      # end
+    end
+
+    it "able to access '/users' page" do
+      get '/users'     
+      last_response.status.must_equal 200
+    end
+
+    it "able to access other user's page" do
+      get "/users/#{@user[:id]}"
+      last_response.status.must_equal 200
+    end
+
+    it "displays all users '/users' page" do
+      number_of_users = @users.count
+      get '/users'
+      users_page = Nokogiri::HTML(last_response.body)
+      li_size = users_page.search("li").size
+      li_size.must_equal number_of_users
+    end
+  end
+
+  #-------------------------------------------------------------------------------------
+  describe "when logged in as a regular user" do
+    before do
+      # DB.transaction(:rollback => :always) do
+        @users.insert(admin: false, fname: "f2", lname: "l2", email: "e2", username: "u2")
+        @users.insert(fname: "different_user", lname: "l2")
+        @regular_user = @users.where(admin: false).first
+        @user = @users.where(fname: "different_user").first
+        env "rack.session", {:current_user_id => @regular_user[:id]}
+      # end
+    end
+
+    it "can view own page" do
+      get "/users/#{@regular_user[:id]}"
+      last_response.status.must_equal 200
+    end
+
+    it "shows 'Access forbidden' when accessing '/users' page" do
+      get '/users'
+      last_response.status.must_equal 403
+    end
+
+    it "shows 'Access forbidden' when accessing other user's page" do
+      get "/users/#{@user[:id]}"
+      last_response.status.must_equal 403
+    end
+  end
+
+  # #-------------------------------------------------------------------------------------
+  describe "testing login, create, update and delete" do
+    before do
+      # DB.transaction(:rollback => :always) do
+        post '/users', params = {fname: "f3", lname: "l3", email: "e3", username: "u3", password: "p3"}
+        @user = @users.where(fname: "f3").first
+      # end
+    end
+
+    # works with DB.transaction
+    it "successfully creates a new user" do
+      @user.wont_be_nil
+    end
+
+    it "successfully logs in a user" do
+      post '/users/login', params = {username: "u3", password: "p3"}
+      follow_redirect!
+      path = last_request.fullpath
+      path.must_equal "/users/#{@user[:id]}"
```

Comment:
Instead of asserting that` session[:current_user_id]` has the correct value or that `ApplicationController#current_user `returns the correct user, I checked that user was redirected to the correct path.

BTW, how do I call(?) `Application#current_user` or any method inside the test? I tried `current_user.wont_be_nil` but it returned a _NameError: undefined local variable or method current_user_. Same with `user_signed_in?.must_equal true`.

Also, wouldn't checking for `session[:current_user_id]` and `Application#current_user` not work inside the test since sessions have to be manually set using `env "rack.session", {:current_user_id => @user[:id]}`?